### ISSUE TITLE
net: regression on native IP stack when an offload driver is present

### ISSUE THREAD

**Issue by RomainPelletant** [zephyrproject-rtos/zephyr#61173]
**Describe the bug**
In case we include ethernet driver (enc424j600) and wifi offload driver (esp-at), the ethernet interface is no more usable and go into hard fault.
We are only using one interface at a time (exchangeable).
But, surprisingly, Wi-Fi interface is working well when selected.

Please also mention any information which could help others to understand
the problem you're facing:
 - The target network interface is always set as default interface

**To Reproduce**
1. include enc424j600 or another native driver
2. include an offloaded driver or esp-at
3. Make something on network (UDP or TCP)

**Expected behavior**
Connect succesfull

** Workaround **
Revert PR #58232

**Impact**
showstopper

**Additional context**
Duplicate of issue #58435

**Comment by rlubos** [comment]
How do you switch between native/offloaded interfaces? I'm asking, because NET_OFFLOAD offloading is known of not being well prepared to work with multiple interfaces in the system (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/net_context.c#L258, https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/net_context.c#L330) - it will always assume the default interface should be used. Are you switching the default interface before allocating socket/net_context?

Anyway, I can see one possible issue with native interfaces when NET_OFFLOAD is enabled - after the change which installs the recv handler before connect, this line might be an issue: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/net_context.c#L2114. In case of native interface, the `iface` pointer on the `context` may still not be set, which would lead to an assert/crash `net_if_is_ip_offloaded()`.

Given that we now assign the interface to the context during context creation for offloaded cases (see https://github.com/zephyrproject-rtos/zephyr/pull/59851) I think we could update `net_if_is_ip_offloaded()` to handle NULL pointer more gracefully (as it would indicate we deal with native interface). Could you check if changing the function as follows helps in your case:
```patch
 static inline bool net_if_is_ip_offloaded(struct net_if *iface)
 {
 #if defined(CONFIG_NET_OFFLOAD)
-       NET_ASSERT(iface);
-       NET_ASSERT(iface->if_dev);
-
-       return (iface->if_dev->offload != NULL);
+       return (iface != NULL && iface->if_dev != NULL &&
+               iface->if_dev->offload != NULL);
 #else
        ARG_UNUSED(iface);
```

**Comment by RomainPelletant** [comment]
Hi @rlubos,

Thank you for your answer.
Before loading drivers, network interface is known is EEPROM and only the right driver is loaded.
It works even when two interfaces drivers are loaded, but before doing network things, I always call **_net_if_set_default();_**

Impressive : I rebased from current 3.4 and apply your patch : it works like a charm.
I looked into socket.c and issue always came around net_if_is_ip_offloaded but I didn't see the inline function in header.

I can do PR and can leave you do so as you want.
Thank you for your help
